pkg/controller/sparkapplication/controller.go

```diff
@@ -501,20 +507,33 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
 func (c *Controller) updateSparkApplicationStatusWithRetries(
 	original *v1alpha1.SparkApplication,
 	updateFunc func(*v1alpha1.SparkApplicationStatus)) *v1alpha1.SparkApplication {
-	toUpdate := original.DeepCopy()
-
 	var lastUpdateErr error
 	for i := 0; i < maximumUpdateRetries; i++ {
-		updated, err := c.tryUpdateStatus(original, toUpdate, updateFunc)
+		toUpdate := original.DeepCopy()
+
+		glog.V(2).Infof("Trying to update SparkApplication %s", original.Name)
+		// Apply update
+		updateFunc(&toUpdate.Status)
+
+		// Let's keep the old App status if this is not a valid transition.
+		if !isValidDriverStateTransition(original.Status.AppState.State, toUpdate.Status.AppState.State) {
+			toUpdate.Status.AppState = original.Status.AppState
```

Comment:
We might need to log a `warning` here with the old and new state.

Reply:
done, just FYI, this will be pretty common since same state transitions are also invalid. e.g. we will log going from RUNNING -> RUNNING.

Reply:
Oh, right, I forgot that transition. Let's add another condition before the check for validity like the following:
```
if toUpdate.Status.AppState.State != original.Status.AppState.State && !isValidDriverStateTransition(original.Status.AppState.State, toUpdate.Status.AppState.State)
```

Reply:
done